src/Adapter/Order/CommandHandler/AddProductToOrderHandler.php

```diff
@@ -202,60 +215,155 @@ private function assertOrderWasNotShipped(Order $order)
      * @param Order $order
      * @param OrderInvoice|null $invoice
      * @param Cart $cart
-     * @param array $productCart
-     *
-     * @return OrderDetail
+     * @param array $cartProducts
      *
      * @throws \PrestaShopDatabaseException
      * @throws \PrestaShopException
      */
-    private function createOrderDetail(Order $order, ?OrderInvoice $invoice, Cart $cart, array $productCart): OrderDetail
+    private function createOrderDetail(Order $order, ?OrderInvoice $invoice, Cart $cart, array $cartProducts): void
     {
         $orderDetail = new OrderDetail();
         $orderDetail->createList(
             $order,
             $cart,
             $order->getCurrentOrderState(),
-            $productCart,
+            $cartProducts,
             !empty($invoice->id) ? $invoice->id : 0
         );
+    }
+
+    /**
+     * When a product is added some other products may be affected (because of CartRule that would
+     * add some gift for example) So we detect which existent products may have been modified in the
+     * cart and update the related OrderDetail
+     *
+     * @param Order $order
+     * @param array $oldProducts
+     * @param array $newProducts
+     * @param Product $product
+     * @param Combination|null $combination
+     *
+     * @throws OrderException
+     * @throws \PrestaShopDatabaseException
+     * @throws \PrestaShopException
+     */
+    private function updateAffectedOrderDetails(
+        Order $order,
+        array $oldProducts,
+        array $newProducts,
+        Product $product,
+        ?Combination $combination
+    ): void {
+        $orderDetails = $order->getOrderDetailList();
+        foreach ($newProducts as $newProduct) {
+            // Created OrderDetail has already been created we don't update it
+            $productMatch = $newProduct['id_product'] == $product->id;
+            $combinationMatch = $combination === null || $newProduct['id_product_attribute'] == $combination->id;
+            if ($productMatch && $combinationMatch) {
+                continue;
+            }
+
+            // New products in the cart have also been created, not need for update
+            $affectedProduct = $this->getMatchingProduct($oldProducts, $newProduct);
+            if (null === $affectedProduct) {
+                continue;
+            }
+
+            $deltaQuantity = (int) $newProduct['cart_quantity'] - (int) $affectedProduct['cart_quantity'];
+            if (0 === $deltaQuantity) {
+                continue;
+            }
+
+            $affectedOrderDetail = null;
+            foreach ($orderDetails as $orderDetailData) {
+                if ((int) $orderDetailData['product_id'] === (int) $affectedProduct['id_product']
```

Comment:
It might be checked in another location in the code but ... could this become a negative integer 😬 ?

Reply:
@matks product ids are always unsigned 🤔

Reply:
Bad line 😅 I wanted to mention `$newUpdatedQuantity = (int) $affectedOrderDetail->product_quantity + $deltaQuantity;`